[Refactoring] Use known alternative in async refactoring if one exists

bnbarham opened the pull request:
Update the async refactorings to use the name from the async alternative
if one is known, rather than always assuming the name matches the
synchronous function.

Note that this could also be used to determine whether results remain
optional or not, but that has been left for a future patch.

Resolves rdar://80612521

bnbarham committed:
[Refactoring] Use known alternative in async refactoring if one exists

Update the async refactorings to use the name from the async alternative
if one is known, rather than always assuming the name matches the
synchronous function.

Note that this could also be used to determine whether results remain
optional or not, but that has been left for a future patch.

Resolves rdar://80612521

bnbarham commented:
@swift-ci please test

bnbarham commented:
Failing test was disabled in another PR that's now merged

@swift-ci please test macOS platform

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -6218,14 +6279,13 @@ class AsyncConverter : private SourceEntityWalker {
   void convertNode(ASTNode Node, SourceLoc StartOverride = {},
-                   bool ConvertCalls = true) {
+                   bool ConvertCalls = true,
+                   bool IncludePreceedingComment = true) {
```
Nit: preceeding -> preceding

bnbarham replied:
Not sure how a double e snuck in there, thanks 🙇

hamishknight reviewed (approved):
LGTM!

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -7357,21 +7450,65 @@ class AsyncConverter : private SourceEntityWalker {
+      addRange(CE->getStartLoc(), CE->getFn()->getEndLoc(),
+               /*ToEndOfToken=*/true);
+    } else {
+      addRange(CE->getStartLoc(), CE->getFn()->getStartLoc(),
+               /*ToEndOfToken=*/false);
```
Yes, not sure what I was thinking there.

In adding a test for that I found another bug where we weren't walking over the DeclRefExpr in a double curry thunk. Also since I just committed the getter fix I'll add a test and fix whatever that breaks as well :)

bnbarham committed:
[Diagnostics] Allow matching accessors and functions for a renamed decl

A lookup on the name provided by `renamed` in `@available` returns the
VarDecl. If the name specified a getter or setter, make sure to grab the
corresponding accessor when comparing candidates.

We currently ignore the basename and parameters specified in the name
for accessors. Checking them would only cause a getter/setter not to
match, there can never be a conflict. Since this is a best effort match
anyway, this seems fine.

bnbarham commented:
@etcwilde I forgot to add tests for renaming to a getter in the attribute tests. 8f4d272 fixes the `RenameDeclRequest` to grab the underlying accessor if the name is `getter:` or `setter:`. Also added a new `printUserFacingName` to `AccessorDecl` so the refactoring code can use it too.

@hamishknight there's now some code to handle the accessor and I changed SemaAnnotator to walk the `DeclRefExpr` inside a double curry thunk rather than just calling `passReference` - the rename would be skipped otherwise.

bnbarham commented:
@swift-ci please test

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -7363,23 +7449,71 @@ class AsyncConverter : private SourceEntityWalker {
+    } else {
+      // Replace the original call with the name of the alternative
+      Identifier Name;
+      if (auto *Accessor = dyn_cast<AccessorDecl>(HandlerDesc.Alternative))
+        Name = Accessor->getStorage()->getBaseIdentifier();
+      else
+        Name = HandlerDesc.Alternative->getBaseIdentifier();
```
Just occurred to me that `getBaseIdentifier` would hit an assertion if this a decl with a special name such as `init`, and I *believe* we support marking those async.

bnbarham replied:
`init` can definitely be marked async, but it wouldn't make sense as a completion handler function since it inherently returns a value (and thus should never be an alternative).

Regardless, it's probably better to check and fallback anyway - only adds a line or two.

hamishknight replied:
Ah right yeah, sounds good 👍

hamishknight reviewed (commented):
Accessor and curry thunk changes look good to me!

bnbarham commented:
Moved the SourceEntityWalker changes to https://github.com/apple/swift/pull/38712. Tests will fail until that's in, so I won't bother running them yet.

bnbarham commented:
@swift-ci please test
